airbyte-integrations/connectors/source-tiktok-marketing/source_tiktok_marketing/streams.py

```diff
@@ -0,0 +1,302 @@
+        )
+        self._max_cursor_date = None
+
+    @property
+    def is_sandbox(self):
+        return self._advertiser_storage.is_sandbox
+
+    @property
+    def advertiser_ids(self) -> List[int]:
+        return self._advertiser_storage.advertiser_ids
+
+    @staticmethod
+    def convert_array_param(arr: List[Union[str, int]]) -> str:
+        return json.dumps(arr)
+
+    def stream_slices(self, **kwargs) -> Iterable[Optional[Mapping[str, Any]]]:
+        """Loads all updated tickets after last stream state"""
+        for advertiser_id in self.advertiser_ids:
+            yield {"advertiser_id": advertiser_id}
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+        stream_slice: Mapping[str, Any] = None,
+        **kwargs
+    ) -> MutableMapping[str, Any]:
+        params = {"page_size": self.page_size}
+        if self.fields:
+            params["fields"] = self.convert_array_param(self.fields)
+        if stream_slice:
+            params.update(stream_slice)
+        return params
+
+
+class IncrementalTiktokStream(FullRefreshTiktokStream, ABC):
+    cursor_field = "modify_time"
+
+    page_size = 1000
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._max_cursor_date = None
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        """All responses have the following pagination data:
+        {
+            "data": {
+                "page_info": {
+                    "total_number": < total_item_count >,
+                    "page": < current_page_number >,
+                    "page_size": < page_size >,
+                    "total_page": < total_page_count >
+                },
+                ...
+           }
+        }
+        """
+
+        page_info = response.json()["data"]["page_info"]
+        if page_info["page"] < page_info["total_page"]:
+            return {"page": page_info["page"] + 1}
+        return None
+
+    def request_params(self, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(next_page_token=next_page_token, **kwargs)
+        if next_page_token:
+            params.update(next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        """Additional data filtering"""
+        state = stream_state.get(self.cursor_field) or self._start_time
+        for record in super().parse_response(response, **kwargs):
+            updated = record[self.cursor_field]
+            if updated <= state:
+                continue
+            elif not self._max_cursor_date or self._max_cursor_date < updated:
+                self._max_cursor_date = updated
+            yield record
```

Comment:
the cdk checkpoints state after each slice ([see here](https://github.com/airbytehq/airbyte/blob/1773e41e47a69d22a1c575bb8a13f34aa4ed06a7/airbyte-integrations/bases/base-python/base_python/cdk/abstract_source.py#L148)) so I'm concerned that because our slices are not time-ordered, we could be missing data by having those slices and then filtering on cursor field on response.

One way around this would be to not use stream_slices but instead override read_records so that we iterate through the advertiser ids in there and add them to request params. This would mean we only checkpoint state at the end of the sync.